File: Introduction_to_MPI/Solutions/c/ex5-collective/collective.c
#include <mpi.h>
#include <stdio.h>

int main (int argc, char* argv[]){

  int rank, size;
  int sum;

  MPI_Init(&argc,&argv);
	
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &size);

  MPI_Allreduce(&rank, &sum, 1, MPI_INT, MPI_SUM, MPI_COMM_WORLD);

  /* Display the result on all ranks, along with the correct answer */
  printf("Rank %2d has sum of ranks %d; Answer %d \n", rank, sum, (size-1)*size/2);

  MPI_Finalize();
	
  return 0;

} 

File: Introduction_to_MPI/Solutions/fortran/ex5-collective/collective.f90
program collective

  use mpi
  implicit none

  integer :: ifail
  integer :: rank, size

  integer :: sum

  call mpi_init(ifail)

  call mpi_comm_rank(MPI_COMM_WORLD, rank, ifail)
  call mpi_comm_size(MPI_COMM_WORLD, size, ifail)

  call mpi_allreduce(rank, sum, 1, MPI_INTEGER, MPI_SUM, MPI_COMM_WORLD, ifail)

  write (unit = *, fmt = '(a,i3,a,i4,a,i4)') &
       "Rank ", rank, " has sum of ranks ", sum, " Answer ", (size - 1)*size/2

  call mpi_finalize(ifail)

end program collective
